sympy/integrals/integrals.py

```diff
@@ -1242,6 +1242,110 @@ def _sage_(self):
+        """
+        from sympy.calculus import singularities
+
+        if len(self.limits) != 1 or len(list(self.limits[0])) != 3:
+            raise ValueError("You need to insert a variable, lower_limit, and upper_limit correctly to calculate "
+                             "cauchy's principal value")
+
+        x, a, b = self.limits[0]
+
+        if not(a.is_comparable and b.is_comparable and a <= b):
+            raise ValueError("The lower_limit must be smaller than or equal to the upper_limit to calculate "
+                             "cauchy's principal value. Also, a and b need to be comparable.")
+
+        if a == b:
+            return 0
+        r = Dummy('r')
+        u = Dummy('u')
+
+        singularities_list = [s for s in singularities(self.function, x) if s.is_comparable and a <= s <= b]
+
+        for i in singularities_list:
+            if ((i == b) or (i == a)):
+                raise ValueError(
+                    'The principal value is not defined in the given interval due to singularity at %d.' % (i))
+
+        f = self.function
+        F = integrate(f, x)
+
+        if F.has(Integral):
+            return self
+        if len(singularities_list) == 0:
+            if a is -oo and b is oo:
+                I = limit(F - F.subs(x, -x), x, oo)
+                return I
+            else:
+                I = limit(F, x, b, '-') - limit(F, x, a, '+')
+                return I
+
+        if len(singularities_list) == 1 and not(a is -oo and b is oo):
+            I = limit(((F.subs(x, singularities_list[0] - r)) - F.subs(x, singularities_list[0] + r)), r, 0,
+                      '+') - limit(F, x, a) + limit(F, x, b)
+            return I
+
+
+        else:
+            if a is -oo and b is oo:
+                Iinf = 0
+                for singular_element in range(len(singularities_list)):
+                    Iinf = Iinf + limit(((F.subs(x, singularities_list[singular_element] - r)) - F.subs(x,
+                                                                                                        singularities_list[
+                                                                                                            singular_element] + r)),
+                                        r, 0, '+') + limit(F.subs(x, u) - F.subs(x, -u), u, oo)
+                return Iinf
+            else:
+                Ib = 0
+                Ia = F.subs(x, singularities_list[0] - r) - limit(F, x, a)
+                for singular_element in range(len(singularities_list)):
+                    Ib = Ib + (F.subs(x, singularities_list[singular_element] - r) - F.subs(x, singularities_list[
+                        singular_element - 1] + r))
+                I1 = limit((limit(F, x, b) - (F.subs(x, singularities_list[len(singularities_list) - 1] + r))) + (
+                        F.subs(x, singularities_list[len(singularities_list) - 1] - r) - F.subs(x, singularities_list[
```

Comment:
These two `F.subs` terms were already added to `Ib` in the for loop above.

Reply:
@jksuom  Yes. You are right. I've addressed that by making the range : `range(1, len(singularities_list) - 1)` instead of `range(1, len(singularities_list))` ( this I just realised ) since just after the `for` loop, the last element is taken care of.